libkbfs/block_journal.go

```diff
@@ -952,9 +978,72 @@ func flushBlockEntries(ctx context.Context, log logger.Logger,
 	return nil
 }
 
+// removeEarliestEntryFromDiskJournal removes the earliest entry from
+// the given disk journal `dj`, which might be different from `j.j`.
+// If `dj` is the same as `j.j`, the entry is also cleared from the
+// in-memory reference map.  If `removeData` is true, the
+// corresponding block data is also removed from disk.
+func (j *blockJournal) removeEarliestEntryFromDiskJournal(
+	entry blockJournalEntry, removeData bool, dj diskJournal) error {
+	earliestOrdinal, err := dj.readEarliestOrdinal()
+	if err != nil {
+		return err
+	}
+
+	_, err = dj.removeEarliest()
+	if err != nil {
+		return err
+	}
+
+	// Remove any of the entry's refs that hasn't been modified by
+	// a subsequent block op (i.e., that has earliestOrdinal as a
+	// tag).
+	isLiveDiskJournal := j.j.dir == dj.dir
+	for id, idContexts := range entry.Contexts {
+		refs := j.refs[id]
+		if len(refs) == 0 && isLiveDiskJournal {
+			continue
+		}
+		for _, context := range idContexts {
+			// Only remove the reference if this is the live disk journal.
+			if isLiveDiskJournal {
+				err := refs.remove(context, earliestOrdinal)
+				if err != nil {
+					return err
+				}
+			}
+			if len(refs) == 0 {
+				delete(j.refs, id)
+
+				if removeData {
+					// Garbage-collect the old entry.  TODO: we'll
+					// eventually need a sweeper to clean up entries left
+					// behind if we crash here.
+					err = j.removeBlockData(id)
+					if err != nil {
+						return err
+					}
+				}
+				break
+			}
+		}
+	}
+	return nil
+}
+
 func (j *blockJournal) removeFlushedEntry(ctx context.Context,
 	ordinal journalOrdinal, entry blockJournalEntry) (
 	flushedBytes int64, err error) {
+	earliestOrdinal, err := j.j.readEarliestOrdinal()
```

Comment:
Even if you put the loop back in this function, I think this check should go first here.